### ISSUE TITLE
Alias the es6 env to es2015

### ISSUE THREAD

**Issue by mgol** [eslint/eslint#7006]
**The version of ESLint you are using.**
3.4.0

**The problem you want to solve.**
ESLint 3.4.0 introduced new preferred numbers for the `ecmaVersion` property, e.g. 2015 instead od 6. The change is not reflected in the `es6` environment name. My current example config is now inconsistent:
```js
module.exports = {
    parserOptions: {
        ecmaVersion: 2016,
    },

    env: {
        es6: true,
    },

    rules: { /* ... */ },
};
```

**Your take on the correct solution to problem.**
An alias env called `es2015` should be created and preferred to the old `es6` one.

**Comment by vitorbal** [comment]
This seems reasonable to me. @eslint/eslint-team any reasons why we wouldn't want to do this that you can think of?

**Comment by platinumazure** [comment]
We should only need to do this no more than twice (es2015 for es6, and es2016 for es7 if that included any new globals), so I'm not worried about an ongoing maintainability cost at this point.

**Comment by nzakas** [comment]
:-1: I don't like having two ways of doing the same thing, and it could lead to a lot of confusion. For instance, what does this mean?

```
env:
  es6: true
  es2015: false
```

It will be hard to keep track of places where you're using one vs. the other, especially if you're using shareable configs that might have one or the other.

I think this is a case where we just say that the way we are doing things is good enough and leave it at that.

**Comment by vitorbal** [comment]
That's a good point. Okay then, given the added complexity that @nzakas pointed out, I would be against this change. Let's see if anyone from the team feels strongly in favor of it.

**Comment by platinumazure** [comment]
Yeah, this is different from ecmaVersion since that's a scalar, not an
object where you can have true/false on multiple envs plus overrides. I
agree, :-1: we shouldn't do this.

**Comment by michaelficarra** [comment]
There's nothing wrong with the ability to write

```yaml
env:
  es6: true
  es2015: false
```

Then the user will just get the error "Both `es6` and `es2015` env options were set to conflicting values. Remove one to continue."

**Comment by ljharb** [comment]
@nzakas it's not good enough long term though - the confusion between "edition number" and "year" basically only applies to ES6 and ES7, since after that people will very likely only use "year". If you don't want to have conflicting options, at the least, it should start using only the year as of ES2017.

**Comment by platinumazure** [comment]
I think we can probably create new envs with the year from now on- @nzakas
was mainly objecting to having synonymous envs. (Correct me if I'm wrong)

**Comment by ljharb** [comment]
@platinumazure the concern then is that you have to then deal with the confusion of "why can i do ES2017 and beyond, but not ES2015 that all these blogs and books refer to"

**Comment by nzakas** [comment]
Yeah, if we need to create new environments going forward, we'll do it in the most logical way possible.

@ljharb you're extrapolating out to a problem we're not talking about here. This issue is only about whether or not to create a second environment that does the same thing as `es6`. If we need to create other environments going forward, we'll name them in the most logical way. My hunch is that we'll probably end up doing something so that people don't have to use both environments and parser options at some point, but we've not yet encountered a problem that's forced us to solve for that.

@michaelficarra the user won't get an error unless we hardcode that equivalence in, and there's no good reason to do that.

**Comment by mgol** [comment]
@nzakas 

I was thinking about what @michaelficarra proposed - basically, to alias `es6` to `es2015` internally so that there is never an `es6` env internally and prevent specifying both envs in one config. That would limit the consequences of such a change to only the config parsing phase.

> I don't like having two ways of doing the same thing

Well, we already have two ways of doing the same thing: `ecmaVersion: 2015` and `ecmaVersion: 6`. This specific duplication was what made me create this issue after all. In a future ESLint major version it'd be possible to remove this duplication by dropping "the old way" of specifying ES versions.

**Comment by nzakas** [comment]
@mgol that would require a special case for this, and I don't see a good reason to do that. 

> Well, we already have two ways of doing the same thing: ecmaVersion: 2015 and ecmaVersion: 6.

Completely different situation. You can only specify one value here, and if the value is 2015 or over, we just subtract 2009 from it. We could have not allowed 2015, but it would have been completely arbitrary. No special casing, no special logic, no need for validation.

**Comment by scriptdaemon** [comment]
@nzakas Consistency is always a good reason.

Regarding:

```
env:
  es6: true
  es2015: false
```

This would be unsupported behavior, so either:

  1. If either is true, then use the environment regardless of what the other option is set to.
  2. If either is false, then don't use the environment regardless of what the other option is set to.
  3. Rely on whatever the config parser parses first.
  4. Rely on whatever the config parser parses last.
  5. Don't care.

And in either case, make sure to document that including both is unsupported behavior.

1 or 2 would probably be the easiest to implement and make the most sense (imo).

**Comment by nzakas** [comment]
TSC Question: should we add an es2015 environment in addition to es6?

**Comment by btmills** [comment]
Based on discussion in the 2016-09-01 TSC meeting, we believe supporting both `es6` and `es2015` in parallel or renaming `es6` to `es2015` would be more confusing  than keeping the name as-is. `es6` will remain, and we will use year-based naming starting with the 2017 edition. (ES7/ES2016 introduced no new globals, so it doesn't have its own environment.)

**Comment by mgol** [comment]
OK. Thank you for considering my proposal.
